### DISCUSSION TITLE
How to guarantee external lib is included by Gulp?

### DISCUSSION THREAD

**Discussion by heberleh** [TIBCOSoftware/spotfire-mods#88] [Q&A] [answered]
I'm trying to use a local lib, that I install with `npm install ../folder/mylib.tgz`, but I get his message:

> No name was provided for external module 'mylib' in output.globals – guessing 'mylib'

In the browser when running the Spotfire Mod then I get:

> Uncaught ReferenceError: mylib is not defined

I tried to include  `external: [ 'mylib' ],`  in the gulpfile.js, but didn't work.
   
I tried to export my lib as `commonjs2` and as `umd`.
I did this for other projects (not Mods), and they worked, but I always use webpack, never used Gulp.

The webpack from `mylib` looks like this:

```js
{
  filename: "index.js",
  library: "mylib",
  libraryTarget: "umd", // or "commonjs2"
}
```

Any ideas?

**Comment by objerke** [comment]
It's not easy to help without further information.

- How is your mod bundled?
- Did you start from any particular mod example?
- How do you reference your external lib from your mod source code?
- Can you use your local lib from any other similar source code project?

I can't think of any of our existing examples that use webpack and gulp together. A mod is in essence only a website with an index.html, its referenced resources, and a manifest. There is nothing special regarding the bundling of resources that I can think of.

**Reply by heberleh** [reply]
I'm using the React  Mod example (from prototype folder) as base, api 1.2.1. I haven't changed the gulp.

I created now another plain React project (`create-react-app`, it probably uses webpack too behind the scenes) and there I import `mylib` and can use it normally, by using `import {myMethod} from "mylib"`.

Typescript actually recognize the library, including the function with types, etc, normally in the React Mod Example.
However, when I try to start it, it complains as I mentioned above. I think that the coding environment recognizes it, but after transpiling with gulp, it is not delivered in the bundle and the browser/spotfire complains that `mylib` is not defined.

Some resources were saying the Gulp for some reason may not recognize the external library, but the current gulpfile.js already has the plugins that they suggest to add.

I'll keep trying. Thanks for your support @objerke

**Reply by heberleh** [reply]
Maybe it has to do with the`tsconfig.json`. In the spotfire mod, it says `"module": "es2015",` while in the react app where the lib is recognized it states `"module": "commonjs",`.
If I change to commonjs, gulp complains.

**Comment by objerke** [comment, marked as answer]
If I remember, the react example was written as a quick one off to show that you can use react in your mod. I think it only can load commonJs from the node_modules folder as [this line](https://github.com/TIBCOSoftware/spotfire-mods/blob/a98d014abaf7ea514a711b0747eb97dbb99fe731/prototypes/dr-spotfire-mods-react-demo-property/gulpfile.js#L59) in the gulpfile suggests. I would look at the react package in `node_modules` and try to mimic the way it exposes its internals. 

![image](https://user-images.githubusercontent.com/8665943/176911244-0c4759e0-7d7c-4edf-9ce9-9fb02152a89a.png)

In general I would not recommend starting new mods from the prototypes folder, as the name suggests they are not polished in any way.
Good luck!

**Reply by heberleh** [reply]
This makes sense! I set my index.js (from mylib), the main file, to be in the root of the module and it worked.
On the other hand, because the types file index.d.ts is still inside another folder, it can't find it. I guess it's for the same reason.

So basically I could have a package.json indicating the folder where index.js is when I use webpack... but with gulp I guess this doesn't work.

Thanks for the help!
